### ISSUE TITLE
gutenberg_locate_block_template slicing template slugs

### ISSUE THREAD

**Issue by shramee** [WordPress/gutenberg#37691]
#### TLDR

Check if specificity `$index` of the located php `$template` is not false before clearing out the template candidates `$templates` in [`gutenberg_locate_block_template`](https://github.com/WordPress/gutenberg/blob/59fbccf780970369682514a6b4e55411e1b7120c/lib/compat/wordpress-5.9/block-template.php#L41)

------------------------------------------

Function `gutenberg_locate_block_template` trims `$templates` (template candidates) even when located php template doesn't match php candidates. This breaks template lookups for custom post types (product in my case) on [lib/compat/wordpress-5.9/block-template.php:61](https://github.com/WordPress/gutenberg/blob/59fbccf780970369682514a6b4e55411e1b7120c/lib/compat/wordpress-5.9/block-template.php#L61).

The comment line says,

```php
// If the template hiearchy algorithm has successfully located a PHP template file,
// we will only consider block templates with higher or equal specificity.
```

for **product** post type `$template` is `./wp-includes/template-canvas.php`, which doesn't seem specific to me at all.

#### Possible Solution

Just check if the `$index` is not false before splicing the array.

### Step-by-step reproduction instructions

1. Create an FSE theme.
2. Add `single-product.html` template.
3. Notice it doesn't work.

### Screenshots, screen recording, code snippet

![image](https://user-images.githubusercontent.com/11048263/147946319-43c1cb1c-2f2d-43f8-854e-e94c28a74b75.png)

### Environment info

Gutenberg Version 12.2.0
WordPress Version 5.8.2

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

No

**Comment by youknowriad** [comment]
Closing as obsolete: the WordPress 5.9 compat shim this report analyzed was removed long ago — template hierarchy resolution lives in core now (locate_block_template), where this code path was rewritten. If the specificity-slicing bug survived into core's implementation, a core Trac ticket against the current function would be the actionable follow-up.
